Document course sync API endpoint

ZacWarham opened the pull request:
Closes #12288 

Given this is the first `POST` endpoint, is it sufficient to just mention it is a POST endpoint as opposed to a separate list for `GET`/`POST` endpoints? Alternatively, could just have `(POST)` next to the name?

reteps commented on docs/api.md:
````diff
@@ -18,6 +18,16 @@ Provide your token via the `Private-Token` header:
+Here is an example of using `curl` for a `POST` request:
+
+```sh
+curl -H "Private-Token: TOKEN" -X POST https://us.prairielearn.com/pl/api/v1/<REST_OF_PATH>
````
I think `-d ""` is a good idea.

ZacWarham replied:
Updated in b3ab388

nwalters512 replied:
Let's use `-X POST` for clarity here.

reteps commented on docs/api.md:
```diff
@@ -18,6 +18,16 @@ Provide your token via the `Private-Token` header:
 
+## Types of Endpoints
+
+When sending a request to an endpoint, it will require you to send either a `GET` or a `POST` request. A `GET` request retrieves information from PrairieLearn, such as gradebook information. Whereas a `POST` request asks PrairieLearn to perform an action. For example, syncing a course repository to PrairieLearn. You will need the correct permissions to be able to perform this request. Unless spec ...
```
```suggestion
Unless specified otherwise, all requests are `GET` requests and will follow the above example format when using `curl`.
```

reteps replied:
I understand this is the exact opposite of what @jonatanschroeder said to do, so maybe this is something to be discussed?

ZacWarham replied:
I do agree that various developers should know this but I don't want to assume all PL instructors are developers

jonatanschroeder replied:
I agree with @ZacWarham, I think it's better to be explicit. Even technically-savvy users may not be familiar with GET vs POST requests, this text clarifies that this is a convention used in PL (even if it is a convention used elsewhere as well). It also provides additional information for instructors testing/debugging their API usage.

reteps reviewed (approved):
Approving with the caveat that @jonatanschroeder and my comments are resolved.

I don't think that a `GET` qualifier next to each endpoint is needed for now.

ZacWarham commented:
Just putting an image here of how the formatting currently looks for indicating it is a `POST` endpoint

![image](https://github.com/user-attachments/assets/b8a60692-7613-4f53-843b-30a4067fd41f)

nwalters512 commented on docs/api.md:
```diff
@@ -18,6 +18,16 @@ Provide your token via the `Private-Token` header:
 
+## Types of Endpoints
+
+When sending a request to an endpoint, it will require you to send either a `GET` or a `POST` request. A `GET` request retrieves information from PrairieLearn, such as gradebook information. Whereas a `POST` request asks PrairieLearn to perform an action. For example, syncing a course repository to PrairieLearn. You will need the correct permissions to be able to perform this request. Unless spec ...
```
Wordsmithing + including links to MDN:

```suggestion
API endpoints require either a [`GET` request](https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Methods/GET) or a [`POST` request](https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Methods/POST). A `GET` request retrieves information from PrairieLearn, such as gradebook information. A `POST` request asks PrairieLearn to perform an action, such as syncing a course GitHub repository. Unless otherwise specified, all requests are `GET` requests and will follow the above example format when using `curl`.
```

If we want to retain a mention of permissions, that should go elsewhere, that has nothing to do with discussions of HTTP methods.

nwalters512 replied:
I think we can safely omit it.

nwalters512 commented on docs/api.md:
```diff
@@ -85,3 +95,11 @@ In the endpoint list below, path components starting with a colon like
   - `/pl/api/v1/course_instances/:course_instance_id/submissions/:submission_id`
   - One specific submission.
+
+- **Course sync: (`POST`)**
+  - `/pl/api/v1/course/:course_id/sync`
+  - Returns the `job_sequence_id` of the course sync job.
```
Nit, I think this is a more conventional way to list a method/URL combination:

```suggestion
- **Course sync:**
  - `POST /pl/api/v1/course/:course_id/sync`
  - Returns the `job_sequence_id` of the course sync job.
```

I'd prefer to explicitly include the method on each endpoint rather than saying "everything is GET by default, though I won't die on this hill.

nwalters512 replied:
That's a good point. I opened https://github.com/PrairieLearn/PrairieLearn/pull/12342 to try to do a more comprehensive rework of how things are displayed, including using a code block to avoid this issue. If you/others like what's going on there, maybe we can merge that first?

nwalters512 replied:
Looks like this has been updated now!

nwalters512 commented on docs/api.md:
````diff
@@ -18,6 +18,16 @@ Provide your token via the `Private-Token` header:
 curl -H "Private-Token: TOKEN" https://us.prairielearn.com/pl/api/v1/<REST_OF_PATH>
 ```
 
+## Types of Endpoints
````
Nit (I don't think of HTTP method as being the "type" of an endpoint):

```suggestion
## Endpoint HTTP methods
```

ZacWarham commented:
Moved in the changes to the formatting from #12342 and removed the clause for 'unless specified'

Just needs a decision for what format to use for the `POST` example and any wording changes from the merge 9e57c91

ZacWarham commented on docs/api.md:
````diff
@@ -12,6 +12,16 @@ Provide your token via the `Private-Token` header:
 curl -H "Private-Token: TOKEN" https://us.prairielearn.com/pl/api/v1/<REST_OF_PATH>
 ```
 
+## Endpoint HTTP methods
+
+API endpoints require either a [`GET` request](https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Methods/GET) or a [`POST` request](https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Methods/POST). A `GET` request retrieves information from PrairieLearn, such as gradebook information. A `POST` request asks PrairieLearn to perform an action, such as syncing a course GitHub repos ...
+
+Here is an example of using `curl` for a `POST` request:
+
+```sh
+curl -H "Private-Token: TOKEN" -X POST https://us.prairielearn.com/pl/api/v1/<REST_OF_PATH>
+```
+
````
We decided not to do this https://github.com/PrairieLearn/PrairieLearn/pull/12318#discussion_r2181101234

nwalters512 commented on docs/api.md:
````diff
@@ -107,3 +117,17 @@ GET /pl/api/v1/course_instances/:course_instance_id/assessment_instances/:assess
+
+### Course sync
+
+#### Start a course sync
+
+```text
+POST /pl/api/v1/course/:course_id/sync
+```
+
+#### Check course sync status
+
+```text
+GET /pl/api/v1/course/:course_id/sync/:job_sequence_id
+```
````
In a previous iteration you had helpful (IMO) text explaining that the first one would return an ID, and that the second one used that ID. Not sure what happened to it, but here's my attempt at adding it back:

````suggestion
#### Start a course sync

```text
POST /pl/api/v1/course/:course_id/sync
```

Returns a `job_sequence_id` that can be used to check on the sync job's status.

#### Check course sync status

```text
GET /pl/api/v1/course/:course_id/sync/:job_sequence_id
```

Returns the status and output of the sync job.
````

ZacWarham replied:
Happy to have it. I just went for a blanket removal copying  #12342 😅

>The descriptions are removed; I felt they were basically redundant given the names for the endpoints themselves.
